### PR TITLE
Introducing a tool to debug the current Zone Stack

### PR DESCRIPTION
**Pull request by samcake**
This pr first goal is to add a way to debug the layering system of the zones happening while the avatar move through the world and enter / leave zones.

Each Zone entity currently has a corresponding RenderItem in the scene which is matching it s bounding box. The idea is to add a RenderJob in the render task which would:
1. pick the visible RenderITems corresponding to the Zones of the current Zone Stack (evaluated in the EntityRenderer)
2. render a bounding box for these, and represent the order (the rank of layering) of the represented zones.

In order to implement a Job doing what was expected i had to introduce the following new concepts to the Scene and Job/Task/Engine

1. Added Selections to the Scene. A Selection is a named list(std::vector) of render items that can be created and updated through the Transactions (just like when updating render items).
The Selections are then available in the rendering jobs for read.
We use this new guy to communicate from the EntityRenderer the current zone stack as an ordered list of the render item IDs of the correspoding Zones of the stack. This is the selection called "RankedZones"

2. The Task class as implemented wasn't letting me expose the input and output just like we can for regular jobs. i adjusted this behavior.
Doing so, i actually removed the need for the user's job class to have to inherit from the class "Task".
The user's job class that want to have child jobs and thus beeing considered a "task" simply need to:
 - declare the inner type JobModel = render::Task::Model<>.
 - Provide a build() method acting as the constructor.
 The Task class is now used explicitly as the root of a hierarchy of Jobs.
The Engine is as such a Task.
Overall this work is improving the inside of these classes and doesn't really change the behavior for the client of these classes, simply add that new support for the input/output that was lacking before.
We use this new property to create the ZoneRendererTask which is the combination of 2 tasks:
 - select the items matching the RankedZones Selection from the list of input Items (the visible opaques).
    The output of that job (SelectSortItems) is sorted according to the order found in the Selection.
    This way the ouput list of zone items is matching the Zone stack order.

 - For these zone items draw their bounding box using the DrawBounds job.
   I adjusted the color behavior so it s using a color wheel /rainbow to go from red to green to blue hense communicating visually the order of the boxes drawn.

3. Simply added a checkbox to my favorite script (Developer/Utilities/Render/debudDeferredLighting.js) that enable / disable that new ZOneRendererTask (off by default ) and then let you visualize the ZOne stack live while navigating through the zones.
 I put together the following Entity pack made of several zones and their matching floor footprint represented as a box.
A bunch of zones (matching the rectangular boxes) defined from this packed entities file:
https://hifi-public.s3.amazonaws.com/sam/entities/BunchOfZones.json
The color of the box matches the the color of the Key light of the zone.
image

## TEST PLAN
- check that enabling ambient occlusion, or shadows through the "developer/Graphics dialog", quitting and restarting recover the same state in which you left it.

- Speed and rendering behavior should be exactly the same in this pr and master.

- In an empty region of a domain, import the "BUnchOFZones.json" entities.
  Enable the ZoneRendererTask. you can this from the script Developer/Utilities/Render/debudDeferredLighting.js, turning the "Zones" checkbox on.
As you walk through the different colored floors, you should see the ZOne stack showing as bounding boxes matching the current state you are in.
The color code is Red = current Zone (always the smallest you are in) then the standard color whell ramp going to green and then blue.
With that setup you also should experience a lighting bug while in the zones, the lighting is constantly flashing while moving or simply rotating. this is a known bug, also present in master.
This work comes as a tool to actually fix that bug next.
Here is what it should look like ith the avatar in the purple zone and the zone stack showing 3 zones. the lighting is purple too.
image

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4a47737d0b11df690e5ca3412325341b0347f3e8 available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-4a47737.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-4a47737.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5d7accca2ba80b9edb0085005f8a39a6047ef70d available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-5d7accc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-5d7accc.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80c781bcdcf38062e842dbde66af861acb4c338c available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-80c781b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-80c781b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1dd98af1ddeb9432f6aecc9590190ba2799feaab available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-1dd98af.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-1dd98af.dmg

File: scripts/developer/utilities/render/deferredLighting.qml [hunk 1]
```diff
@@ -189,6 +189,11 @@ Column {
             checked: Render.getConfig("DrawOverlayTransparentBounds")["enabled"]
             onCheckedChanged: { Render.getConfig("DrawOverlayTransparentBounds")["enabled"] = checked }
         }
+        CheckBox {
+            text: "Zones"
+            checked: Render.getConfig("DrawZones")["enabled"]
+            onCheckedChanged: { Render.getConfig("ZoneRenderer")["enabled"] = checked; Render.getConfig("DrawZones")["enabled"] = checked; }
+        }
```
Comment (samcake) [inline scripts/developer/utilities/render/deferredLighting.qml]: Adding this simple ui to represent the bounding box of the zone stack

File: scripts/developer/utilities/render/debugDeferredLighting.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ var qml = Script.resolvePath('deferredLighting.qml');
 var window = new OverlayWindow({
     title: 'Lighting',
     source: qml,
-    width: 400, height:280,
+    width: 400, height:350,
```
Comment (samcake) [inline scripts/developer/utilities/render/debugDeferredLighting.js]: Need more height for that debug window

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 3]
```diff
@@ -266,6 +268,9 @@ bool EntityTreeRenderer::findBestZoneAndMaybeContainingEntities(QVector<EntityIt
             }
         }
         _layeredZones.apply();
+
+        applyLayeredZones();
```
Comment (samcake) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: AS we are introducing the zones in the render Scene, we need this other call to propagate the new current ZOne Stack to the scene through a Selection

File: libraries/render/src/render/CullTask.h [hunk 4]
```diff
@@ -109,85 +109,6 @@ namespace render {
         void run(const SceneContextPointer& sceneContext, const RenderContextPointer& renderContext, const ItemSpatialTree::ItemSelection& inSelection, ItemBounds& outItems);
     };
 
-    class FilterItemSelectionConfig : public Job::Config {
```
Comment (samcake) [inline libraries/render/src/render/CullTask.h]: All these classes which were in  "CullTask" are now moving to their own file "FilterTask"

File: libraries/render/src/render/FilterTask.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
+//
```
Comment (samcake) [inline libraries/render/src/render/FilterTask.cpp]: All these classes which were in  "CullTask" are now moving to their own file here "FilterTask"

File: libraries/render/src/render/FilterTask.h [hunk 6]
```diff
@@ -0,0 +1,147 @@
+//
```
Comment (samcake) [inline libraries/render/src/render/FilterTask.h]: All these classes which were in "CullTask" are now moving to their own file here "FilterTask"

Comment (samcake) [inline libraries/render/src/render/FilterTask.h]: Also tried to clean up the naming scheme as much as possible

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79bcf957927935d3d21498e17c2bea6a8aeff666 available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-79bcf95.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-79bcf95.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 612acbc6d5fab07c7fb756c320fe8d147e0c3ed6 available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-612acbc.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-612acbc.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5daf365784465c376a3702e9a04be6a125d4d873 available at: https://gist.github.com/09ac7e2548de63780a9777d29afa4862<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-5daf365.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10190-5daf365.dmg

**Comment by Nex-Pro** [comment]
Looking at this.

**Comment by Nex-Pro** [comment]
### I think this is a almost a pass.

Only real issue noticed is that the zone highlighting is very dependent on proximity.
See screen grabs.

You will see I moved about one meter forward and other zone bounds became visible.

hifi-snap-by--on-2017-04-12_18-58-41

hifi-snap-by--on-2017-04-12_18-58-53


After some initial issues trying to determine if the correct zone was active I threw down a set of grey cubes in each zone and overlap. This made checking easier.



Ambient occlusion and shadow states are preserved across restarts.

Zone checkbox works correctly in the debugDeferredLighting dialog.

**Comment by samcake** [comment]
@Nex-Pro Thanks Piper,
Yes the position of the avatar is what determine in which zone we are in.
If you use 3rd person camera as far back as possible from the avatar, you see the logic as you move through the zones. it s basically the center of the hips that triggers it.

THanks for the test !

**Pull request merged**